esphome/components/remote_base/abbwelcome_protocol.cpp

```diff
@@ -0,0 +1,123 @@
+#include "abbwelcome_protocol.h"
+#include "esphome/core/log.h"
+
+namespace esphome {
+namespace remote_base {
+
+static const char *const TAG = "remote.abbwelcome";
+
+static const uint32_t BIT_ONE_SPACE_US = 102;
+static const uint32_t BIT_ZERO_MARK_US = 32;  // 18-44
+static const uint32_t BIT_ZERO_SPACE_US = BIT_ONE_SPACE_US - BIT_ZERO_MARK_US;
+static const uint16_t BYTE_SPACE_US = 210;
+
+uint8_t ABBWelcomeData::calc_cs_() const {
+  uint8_t checksum = 0;
+  for (uint8_t i = 0; i < this->size() - 1; i++) {
+    uint16_t temp = checksum ^ (this->data_[i]);
+    temp = temp ^ (uint16_t) (((uint32_t) temp << 0x11) >> 0x10) ^ (uint16_t) (((uint32_t) temp << 0x12) >> 0x10) ^
+           (uint16_t) (((uint32_t) temp << 0x13) >> 0x10) ^ (uint16_t) (((uint32_t) temp << 0x14) >> 0x10) ^
+           (uint16_t) (((uint32_t) temp << 0x15) >> 0x10) ^ (uint16_t) (((uint32_t) temp << 0x16) >> 0x10) ^
+           (uint16_t) (((uint32_t) temp << 0x17) >> 0x10);
+    checksum = (temp & 0xfe) ^ ((temp >> 8) & 1);
+  }
+  return ~checksum;
+}
+
+void ABBWelcomeProtocol::encode_byte_(RemoteTransmitData *dst, uint8_t data) const {
+  // space = bus high, mark = activate bus pulldown
+  dst->mark(BIT_ZERO_MARK_US);
+  uint32_t next_space = BIT_ZERO_SPACE_US;
+  for (uint8_t mask = 1 << 7; mask; mask >>= 1) {
+    if (data & mask) {
+      next_space += BIT_ONE_SPACE_US;
+    } else {
+      dst->space(next_space);
+      dst->mark(BIT_ZERO_MARK_US);
+      next_space = BIT_ZERO_SPACE_US;
+    }
+  }
+  next_space += BYTE_SPACE_US;
+  dst->space(next_space);
+}
+
+void ABBWelcomeProtocol::encode(RemoteTransmitData *dst, const ABBWelcomeData &src) {
+  dst->set_carrier_frequency(0);
+  uint32_t reserve_count = 0;
+  for (unsigned i = 0; i < src.size(); i++) {
+    reserve_count += 2 * (9 - (src[i] & 1) - ((src[i] >> 1) & 1) - ((src[i] >> 2) & 1) - ((src[i] >> 3) & 1) -
+                          ((src[i] >> 4) & 1) - ((src[i] >> 5) & 1) - ((src[i] >> 6) & 1) - ((src[i] >> 7) & 1));
+  }
+  dst->reserve(reserve_count);
+  for (unsigned i = 0; i < src.size(); i++)
+    encode_byte_(dst, src[i]);
```

Comment:
Please make sure to prefix all class members and member functions with `this->`
```suggestion
    this->encode_byte_(dst, src[i]);
```

Reply:
Thanks for the review! I try to add `this->` to all class members but seems I missed a few places. Do you know if there is an automated method to find the places where `this->` is missing?

Reply:
I don't know of any automated way sorry.